cds/modules/records/permissions.py

```diff
@@ -0,0 +1,309 @@
+        elif action in cls.admin_actions:
+            return cls(record, has_admin_permission)
+        else:
+            return cls(record, deny)
+
+
+class RecordPermission(object):
+    """Record permission.
+
+    - Create action given to any authenticated user.
+    - Read access given to everyone.
+    - Update access given to record owners.
+    - Delete access given to admins only.
+    """
+
+    create_actions = ['create']
+    read_actions = ['read']
+    read_files_actions = ['read-files']
+    update_actions = ['update']
+    delete_actions = ['delete']
+
+    def __init__(self, record, func, user):
+        """Initialize a file permission object."""
+        self.record = record
+        self.func = func
+        self.user = user or current_user
+
+    def can(self):
+        """Determine access."""
+        return self.func(self.user, self.record)
+
+    @classmethod
+    def create(cls, record, action, user=None):
+        """Create a record permission."""
+        if action in cls.create_actions:
+            return cls(record, allow, user)
+        elif action in cls.read_actions:
+            return cls(record, has_read_record_permission, user)
+        elif action in cls.read_files_actions:
+            return cls(record, has_read_files_permission, user)
+        elif action in cls.update_actions:
+            return cls(record, has_update_permission, user)
+        elif action in cls.delete_actions:
+            return cls(record, has_admin_permission, user)
+        else:
+            return cls(record, deny, user)
+
+
+class DepositPermission(RecordPermission):
+    """Deposit permission.
+
+    - Read action given to record owners.
+    - Delete action given to record owners (still subject to being unpublished)
+    """
+
+    @classmethod
+    def create(cls, record, action, user=None):
+        """Create a deposit permission."""
+        if action in cls.read_actions:
+            return cls(record, has_update_permission, user)
+        elif action in cls.delete_actions:
+            return cls(record, has_update_permission, user)
+        return super(DepositPermission, cls).create(record, action, user=user)
+
+
+#
+# Utility functions
+#
+def deny(user, record):
+    """Deny access."""
+    return False
+
+
+def allow(user, record):
+    """Allow access."""
+    return True
+
+
+def has_read_files_permission(user, record):
+    """Check if user has read access to the record's files."""
+    # Allow everyone for now
+    # TODO: decide on files access rights
+    return True
+
+
+def has_read_record_permission(user, record):
+    """Check if user has read access to the record."""
+    # Allow everyone for public records
+    if '_access' not in record:
+        return True
+
+    # Allow e-group members
+    user_groups = get_user_provides()
+    read_access_groups = record['_access']['read']
+    if not read_access_groups:
+        return True
+    if set(user_groups).intersection(set(read_access_groups)):
```

Comment:
Use `set1.isdisjoint(set2)` instead, as you do not really need all common elements but just the existence of one, which is faster.